internal/collector/processor/nodebatcher/node_batcher.go

```diff
@@ -0,0 +1,467 @@
+	"go.uber.org/zap"
+
+	"github.com/census-instrumentation/opencensus-service/internal/collector/processor"
+)
+
+const (
+	initBatchCap             = uint32(1024)
+	nodeDead                 = uint32(1)
+	batchClosed              = uint32(1)
+	tickerPendingNodesBuffer = 16
+
+	defaultRemoveAfterCycles = uint32(10)
+	defaultSendBatchSize     = uint32(8192)
+	defaultNumTickers        = 4
+	defaultTickTime          = 1 * time.Second
+	defaultTimeout           = 1 * time.Second
+)
+
+// batcher is a component that accepts spans, and places them into batches grouped by node and resource.
+//
+// batcher implements processor.SpanProcessor
+//
+// batcher is a composition of four main pieces. First is its buckets map which maps nodes to buckets.
+// Second is the nodeBatcher which keeps a batch associated with a single node, and sends it downstream.
+// Third is a bucketTicker that ticks every so often and closes any open and not recently sent batches.
+// Fourth is a batch which accepts spans and return when it should be closed due to size.
+//
+// When we no longer have to batch by node, the following changes should be made:
+//   1) batcher should be removed and nodeBatcher should be promoted to batcher
+//   2) bucketTicker should be simplified significantly and replaced with a single ticker, since
+//      tracking by node is no longer needed.
+type batcher struct {
+	buckets sync.Map
+	sender  processor.SpanProcessor
+	tickers []*bucketTicker
+	name    string
+	logger  *zap.Logger
+
+	removeAfterCycles uint32
+	sendBatchSize     uint32
+	numTickers        int
+	tickTime          time.Duration
+	timeout           time.Duration
+
+	bucketMu sync.RWMutex
+}
+
+var _ processor.SpanProcessor = (*batcher)(nil)
+
+// NewBatcher creates a new batcher that batches spans by node and resource
+func NewBatcher(name string, logger *zap.Logger, sender processor.SpanProcessor, opts ...Option) processor.SpanProcessor {
+	// Init with defaults
+	b := &batcher{
+		name:   name,
+		sender: sender,
+		logger: logger,
+
+		removeAfterCycles: defaultRemoveAfterCycles,
+		sendBatchSize:     defaultSendBatchSize,
+		numTickers:        defaultNumTickers,
+		tickTime:          defaultTickTime,
+		timeout:           defaultTimeout,
+	}
+
+	// Override with options
+	for _, opt := range opts {
+		opt(b)
+	}
+
+	// start tickers after options loaded in
+	b.tickers = newStartedBucketTickersForBatch(b)
+	return b
+}
+
+// ProcessSpans implements batcher as a SpanProcessor and takes the provided spans and adds them to
+// batches
+func (b *batcher) ProcessSpans(request *agenttracepb.ExportTraceServiceRequest, spanFormat string) (uint64, error) {
+	bucketID := b.genBucketID(request.Node, request.Resource, spanFormat)
+	bucket := b.getOrAddBucket(bucketID, request.Node, request.Resource, spanFormat)
+	bucket.add(request.Spans)
+	return 0, nil
+}
+
+func (b *batcher) genBucketID(node *commonpb.Node, resource *resourcepb.Resource, spanFormat string) string {
+	h := md5.New()
```

Comment:
I'm curious about the numbers that you get in `BenchmarkGenBucketID`

Reply:
```
BenchmarkGenBucketID/composite-md5-bigInput-12         	  500000	      3229 ns/op
BenchmarkGenBucketID/composite-md5-smallInput-12       	 3000000	       517 ns/op
```

Reply:
The result for smallinput is good (comparable to other small functions that we have), the biginput doesn't seem worth any investment at this point since the code is going to change soon when we fully switch to resource instead of node.

Reply:
@pjanotti how are you getting those comparisons of worthiness? Seems like we don't yet have a reference benchmark to gauge against. I made a suggestion to benchmark perhaps stringifying vs md5 hashes.

Reply:
I think both are important -- the proto marshalling actually does equally well with the big and small -- its perf difference is the same multiple (2-3x) for both big and small vs stringify

Reply:
@odeke-em yeah, we should add some more benchmarks and publish them as reference. I currently get ~63K for unmarshaling a zipkinv1 JSON batch , and around ~10Kns for a similar batch in zipkinv1 thrift. For comparison that [tiny optimization that was merged recently](https://github.com/census-instrumentation/opencensus-service/pull/360) had the function with worst case before optimization at around 360ns.